### PR TITLE
i#7050: Remove preempted and faulting instruction from the drmemtraces.

### LINKED ISSUE #7050: tool.drcacheoff.invariant_checker failed with "Missing read/write records"
When a instruction is preempted with no or incomplete read/write records, the invariant checker reports "Missing read/write records" error.

For example:

    ifetch       3 byte(s) @ 0x0000000010f94627 d1 60 20             shl    $0x00000001 0x20(%rax)[4byte] -> 0x20(%rax)[4byte]
    read         4 byte(s) @ 0x000015556606b238 by PC 0x0000000010f94627
    <marker: kernel xfer from 0x10f94627 to handler>

The shl instruction should be followed by a read and a write record. It was preempted in the above example and
the invariant checker reported an error.

The fix is to reset the expected read and write record counts when a kernel transfer event occurs.

### PR DESCRIPTION
**Pull request by ivankyluk**
Change the implementation to remove preempted and faulting instructions from drmemtraces. The objective of the change is to remove non-retired instructions and the corresponding memrefs from drmemtraces.

When an instruction is removed, a new marker TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION will be added. Its value is the encoding of the removed instruction up to the size of a pointer.

Renamed handle_kernel_interrupt_and_markers() to handle_rseq_abort_marker() since the function has been modified to handle rseq abort.

In oder to remove preempted instructions and memrefs, a new function preempted_by_kernel_event() is added to look for KERNEL EVENT marker which may be preceded by memrefs. If a KERNEL EVENT marker is found with the same PC, the instruction and any following memrefs are removed.

Add unit tests to cover instruction and memref removed caused by a KERNEL EVENT.

Update offline-legacy-int-offs.templatex, offline-burst_aarch64_sys.templatex and signal_invariants.c to account for removed instructions.

Fixes #7050

### CODE REVIEW

File: clients/drcachesim/tests/offline-burst_aarch64_sys.templatex [hunk 1]
```diff
@@ -13,7 +13,7 @@ Total counts:
      .* total data loads
```
Comment (derekbruening) [inline clients/drcachesim/tests/offline-burst_aarch64_sys.templatex:13]: I looked back at the details and I really do not understand what is happening with this.  Could you explain how this scenario occurred?  I wrote some of my confusion at https://github.com/DynamoRIO/dynamorio/issues/7050#issuecomment-2443133453.

Putting aside the fact that a regular asynch signal should not cause this kind of thing: if there were some real preempt from say thread relocation, why isn't raw2trace filling in the rest of the instructions in the block?  How is the handler code running already?  That makes it sound like raw2trace is already truncating the rest of the block and somehow solving #5790?  On that note is this PR as written solving #5790?

Comment (derekbruening) [reply clients/drcachesim/tests/offline-burst_aarch64_sys.templatex:13]: What is in the raw trace records?  What is the instruction count for the `shl` block?  What is the exact raw trace order of the branch's block PC, shl block PC, shl address, and signal marker?

Comment (ivankyluk) [reply clients/drcachesim/tests/offline-burst_aarch64_sys.templatex:13]: Here's the trace without the change:
```
        1436         982:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534568 54000061   b.ne   $0x0000aaaae1534574 (untaken)
        1437         983:       94080 ifetch       4 byte(s) @ 0x0000aaaae153456c 97ffffbd   bl     $0x0000aaaae1534460 -> %x30
        1438         984:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534460 a9be7bfd   stp    %x29 %x30 %sp $0xffffffffffffffe0 -> -0x20(%sp)[16byte] %sp
        1439         984:       94080 write       16 byte(s) @ 0x0000fffffb68e260 by PC 0x0000aaaae1534460
        1440         985:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534464 910003fd   add    %sp $0x0000 lsl $0x00 -> %x29
        1441         986:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534468 90001bc0   adrp   <rel> 0x0000aaaae18ac000 -> %x0
        1442         987:       94080 ifetch       4 byte(s) @ 0x0000aaaae153446c f944dc00   ldr    +0x09b8(%x0)[8byte] -> %x0
        1443         987:       94080 read         8 byte(s) @ 0x0000aaaae18ac9b8 by PC 0x0000aaaae153446c
        1444         988:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534470 f9400001   ldr    (%x0)[8byte] -> %x1
        1445         988:       94080 read         8 byte(s) @ 0x0000fffeddbabb58 by PC 0x0000aaaae1534470
        1446         989:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534474 f9000fe1   str    %x1 -> +0x18(%sp)[8byte]
        1447         989:       94080 write        8 byte(s) @ 0x0000fffffb68e278 by PC 0x0000aaaae1534474
        1448         990:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534478 d2800001   movz   $0x0000 lsl $0x00 -> %x1
        1449         991:       94080 ifetch       4 byte(s) @ 0x0000aaaae153447c b90017ff   str    %wzr -> +0x14(%sp)[4byte]
        1450         991:       94080 write        4 byte(s) @ 0x0000fffffb68e274 by PC 0x0000aaaae153447c
        1451         992:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534480 910053e0   add    %sp $0x0014 lsl $0x00 -> %x0
        1452         993:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534484 d5087620   dc_ivac (%x0)[1byte]
        1453         993:       94080 dflush      64 byte(s) @ 0x0000fffffb68e274 by PC 0x0000aaaad5087620
        1454         993:       94080 <marker: kernel xfer from 0xaaaae1534484 to handler>
        1455         993:       94080 <marker: signal #4>
        1456         993:       94080 <marker: timestamp 13374886167543675>
        1457         993:       94080 <marker: tid 94080 on core 53>
        1458         994:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534384 a9be7bfd   stp    %x29 %x30 %sp $0xffffffffffffffe0 -> -0x20(%sp)[16byte] %sp
        1459         994:       94080 write       16 byte(s) @ 0x0000fffffb68cff0 by PC 0x0000aaaae1534384
```

Raw trace:

```
0010c0 0000fffffb68e29c 0000fffffb68e29c
0010d0 200200000006456c 2024000000064460 1437         983:       94080 ifetch       4 byte(s) @ 0x0000aaaae153456c 97ffffbd bl     $0x0000aaaae1534460 -> %x30
                                         1438         984:       94080 ifetch       4 byte(s) @ 0x0000aaaae1534460 a9be7bfd stp    %x29 %x30 %sp $0xffffffffffffffe0 -> -0x20(%sp)[16byte] %sp
0010e0 0000fffffb68e260 0000aaaae18ac000
0010f0 0000fffeddbabb58 0000fffffb68e278
001100 0000fffffb68e274 0000fffffb68e274 1450         991:       94080 write        4 byte(s) @ 0x0000fffffb68e274 by PC 0x0000aaaae153447c
                                         1450         991:       94080 write        4 byte(s) @ 0x0000fffffb68e274 by PC 0x0000aaaae153447c
001110 c200aaaae1534484 c228000000000004 1454         993:       94080 <marker: kernel xfer from 0xaaaae1534484 to handler>
                                         1455         993:       94080 <marker: signal #4>
001120 802f84635aed7f7b c203000000000035
001130 2022000000064384 0000fffffb68cff0
001140 0000fffffb68d00c 0000aaaae18bde60
001150 0000aaaae18bde60 0000aaaae18bde60
```

Comment (derekbruening) [reply clients/drcachesim/tests/offline-burst_aarch64_sys.templatex:13]: The plan is:

- File a separate bug on understanding why an asynchronous signal is delivered mid-block; that won't block changing the raw2trace behavior
- We decided we do want to roll back on asynchronous signals
- The only question is for synchronous signals do we want to roll back and also insert some kind of marker: since there are cases like INT3 where the continuation PC is the next PC so you'd never see the INT3 PC.

Comment (derekbruening) [reply clients/drcachesim/tests/offline-burst_aarch64_sys.templatex:13]: I feel like we have not 100% figured out what we want to do yet, so I don't feel informed enough to review this yet.

As noted above, what is the plan for synchronous signals?

Also, do we want to first understand why the asynch are delivered mid-block: if that turns out to be some problematic issue in DR and we change it, would that affect the decisions here?

Comment (ivankyluk) [reply clients/drcachesim/tests/offline-burst_aarch64_sys.templatex:13]: Let me investigate why the async signals are delivered mid-block,

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 2]
```diff
@@ -1816,6 +1816,59 @@ raw2trace_t::analyze_elidable_addresses(raw2trace_thread_data_t *tdata, uint64 m
...
+            continue;
+        }
+        if (next_entry->extended.ext == OFFLINE_EXT_TYPE_MARKER &&
+            next_entry->extended.valueB == TRACE_MARKER_TYPE_KERNEL_EVENT) {
+            uintptr_t marker_val = 0;
+            if (!get_marker_value(tdata, &next_entry, &marker_val)) {
+                return false;
+            }
+            // A signal/exception marker in the next entry could be at any point
+            // among non-memref instrs, or it could be after this bb.
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:1849]: Why "non-memref instrs"?  It could happen on a load/store instr.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 3]
```diff
@@ -1816,6 +1816,59 @@ raw2trace_t::analyze_elidable_addresses(raw2trace_thread_data_t *tdata, uint64 m
...
+                // We have only the offs, so we can't handle differing modules for
+                // the source and target for legacy traces.
+                if (marker_val == cur_offs)
+                    is_preempted = true;
+            } else {
+                if (marker_val == cur_pc)
+                    is_preempted = true;
+            }
+        }
+        break;
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:1862]: What about TRACE_MARKER_TYPE_PHYSICAL_ADDRESS or related markers: can those appear mid-block?

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 4]
```diff
@@ -1816,6 +1816,59 @@ raw2trace_t::analyze_elidable_addresses(raw2trace_thread_data_t *tdata, uint64 m
...
+    }
+    // Move the entries back to the pre_read queues.
+    while (!pre_read_buffer.empty()) {
+        tdata->pre_read.push_front(pre_read_buffer.front());
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:1866]: Isn't this reversing the order?
This should use queue_entry() instead, right?

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 5]
```diff
@@ -1825,12 +1878,13 @@ raw2trace_t::process_memref(raw2trace_thread_data_t *tdata, trace_entry_t **buf_
...
                             DR_PARAM_OUT bool *interrupted)
 {
-    if (!append_memref(tdata, buf_in, instr, memref, write, reg_vals,
-                       reached_end_of_memrefs))
-        return false;
-    return handle_kernel_interrupt_and_markers(tdata, buf_in, cur_pc, cur_offs,
-                                               instr->length(), instrs_are_separate,
-                                               interrupted);
+    if (*interrupted) {
+        // Flush the memref entry.
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:1882]: Explain that we do not try to indicate which memref might have caused a fault: we just omit them all, along with the instruction fetch.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 6]
```diff
@@ -2023,38 +2077,44 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
...
-            // rely on our own reader.
-            if (!was_prev_instr_rep_string(tdata)) {
-                set_prev_instr_rep_string(tdata, true);
-                buf->type = TRACE_TYPE_INSTR;
-            } else {
-                log(3, "Skipping instr fetch for " PFX "\n", (ptr_uint_t)decode_pc);
-                // We still include the instr to make it easier for core simulators
-                // (i#2051).
-                buf->type = TRACE_TYPE_INSTR_NO_FETCH;
+        // i#7050: remove instructions which were preempted by a kernel event.
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:2080]: Please update as this includes faults which are not "preempted".

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 7]
```diff
@@ -2023,38 +2077,44 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
...
-        if (!skip_icache && record_encoding_emitted(tdata, decode_pc)) {
-            if (!append_encoding(tdata, decode_pc, instr->length(), buf, buf_start))
-                return false;
-        }
-
-        // XXX i#1729: make bundles via lazy accum until hit memref/end, if
-        // we don't need encodings.
-        buf->type = instr->type();
-        if (buf->type == TRACE_TYPE_INSTR_MAYBE_FETCH) {
-            // We want it to look like the original rep string, with just one instr
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:2035]: It may be best to have a rep string fault test to ensure it rolls back to the right point.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 8]
```diff
@@ -2064,20 +2124,7 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
...
-        bool interrupted = false;
-        if (!handle_kernel_interrupt_and_markers(tdata, &buf, cur_pc, cur_offs,
-                                                 instr->length(), instrs_are_separate,
-                                                 &interrupted))
-            return false;
-        if (interrupted) {
-            log(3, "Stopping bb at kernel interruption point %p\n", cur_pc);
-        }
-        // We need to interleave instrs with memrefs.
-        // There is no following memref for (instrs_are_separate && !skip_icache).
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:2079]: Please keep these 2 comment lines.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 9]
```diff
@@ -2104,32 +2151,52 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
...
+                    if (preempted)
+                        break;
+                }
+            }
+        } else {
+            // Flush memref entries.
+            if (preempted) {
+                const offline_entry_t *next_entry = get_next_entry(tdata);
+                while (next_entry != nullptr) {
+                    if (next_entry->addr.type != OFFLINE_TYPE_MEMREF &&
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:2184]: Move this into the while() condition?  That should make for simpler code.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 10]
```diff
@@ -2104,32 +2151,52 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
...
+                        next_entry->addr.type != OFFLINE_TYPE_MEMREF_HIGH)
                         break;
+                    next_entry = get_next_entry(tdata);
                 }
+                if (next_entry != nullptr)
+                    tdata->pre_read.push_front(*next_entry);
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:2190]: Call unread_last_entry() instead?

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 11]
```diff
@@ -2104,32 +2151,52 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
...
-                                        nullptr, &interrupted))
+                                        nullptr, &preempted))
                         return false;
-                    if (interrupted)
+                    if (preempted)
+                        break;
+                }
+            }
+        } else {
+            // Flush memref entries.
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:2180]: But if we have this explicit loop, why do we need the signal interruption check inside process_memref()?  Won't that never be reached?

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 12]
```diff
@@ -2104,32 +2151,52 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
...
+                    if (next_entry->addr.type != OFFLINE_TYPE_MEMREF &&
+                        next_entry->addr.type != OFFLINE_TYPE_MEMREF_HIGH)
                         break;
+                    next_entry = get_next_entry(tdata);
                 }
+                if (next_entry != nullptr)
+                    tdata->pre_read.push_front(*next_entry);
             }
         }
+        // Check for a signal *after* the instruction.
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:2193]: Why do we need this if we're now checking before each instruction?

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 13]
```diff
@@ -2104,32 +2151,52 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
...
+        bool interrupted = false;
+        if (!handle_kernel_interrupt_and_markers(tdata, &buf, cur_pc, cur_offs,
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:2195]: Can we now delete the handle_kernel_interrupt_and_markers() function?  Doesn't the pre-instr "preempt" check subsume it?  If it's only needed for rseq: please rename and simplify it.

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp [hunk 14]
```diff
@@ -3156,6 +3156,248 @@ test_ifiltered(void *drcontext)
...
+#endif
+             check_entry(entries, idx, TRACE_TYPE_INSTR, -1, offs_load) &&
+             check_entry(entries, idx, TRACE_TYPE_READ, -1) &&
+             // The store instruction and the memref are removed.
+             check_entry(entries, idx, TRACE_TYPE_MARKER,
+                         TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION) &&
+             check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_KERNEL_EVENT,
+                         offs_store) &&
+             check_entry(entries, idx, TRACE_TYPE_ENCODING, -1) &&
+             check_entry(entries, idx, TRACE_TYPE_INSTR, -1, offs_move) &&
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_unit_tests.cpp:3288]: Wait: why is the move here?  Wouldn't we either go to a signal handler or exit if no handler?  Why is this different from the prior case?

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp [hunk 15]
```diff
@@ -3156,6 +3156,248 @@ test_ifiltered(void *drcontext)
...
+            check_entry(entries, idx, TRACE_TYPE_THREAD, -1) &&
+            check_entry(entries, idx, TRACE_TYPE_PID, -1) &&
+            check_entry(entries, idx, TRACE_TYPE_MARKER,
+                        TRACE_MARKER_TYPE_CACHE_LINE_SIZE) &&
+            check_entry(entries, idx, TRACE_TYPE_MARKER,
+                        TRACE_MARKER_TYPE_CHUNK_INSTR_COUNT) &&
+            check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP) &&
+            check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID) &&
+            // The rep_stos instruction and the write record are removed because of the
+            // asynchronous signal.
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_unit_tests.cpp:3389]: The rep stos hits corner cases when it's repeated with "unfetched" records on the subsequent iterations.  I would think that's what we'd want to test: the signal arriving on a later iteration.

File: clients/drcachesim/tests/signal_invariants.c [hunk 16]
```diff
@@ -190,8 +190,10 @@ GLOBAL_LABEL(FUNCNAME:)
 #define FUNCNAME test_signal_midbb
         DECLARE_FUNC(FUNCNAME)
 GLOBAL_LABEL(FUNCNAME:)
-        /* prefetcht2's address is the instr count until a signal */
-        prefetcht2 [3]
+        /* prefetcht2's address is the instr count until a signal not counting
+         * the faulting instr.
```
Comment (derekbruening) [inline clients/drcachesim/tests/signal_invariants.c:194]: Add "because the faulting instr is removed from the trace"?  Ditto below.

File: clients/drcachesim/tools/view.cpp [hunk 17]
```diff
@@ -493,6 +493,10 @@ view_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
             std::cerr << "<marker: vector length " << memref.marker.marker_value
                       << " bytes>\n";
             break;
+        case TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION:
+            std::cerr << "<marker: uncompleted instruction, encoding 0x" << std::hex
+                      << memref.marker.marker_value << std::dec << ">\n";
```
Comment (derekbruening) [inline clients/drcachesim/tools/view.cpp:498]: Is it worth trying to print in the encoding style of the ISA (separate bytes for x86, eg..)?  Since it may be incomplete: maybe not.  Yeah I would just leave it like this.  How about adding a comment though?

File: clients/drcachesim/tracer/raw2trace.h [hunk 18]
```diff
@@ -1488,19 +1488,18 @@ class raw2trace_t {
...
+    // memrefs to search for a kernel event marker. If the kernel event marker has
+    // the value of cur_pc, it returns true.
     bool
-    process_memref(raw2trace_thread_data_t *tdata, trace_entry_t **buf_in,
-                   const instr_summary_t *instr, instr_summary_t::memref_summary_t memref,
-                   bool write, std::unordered_map<reg_id_t, addr_t> &reg_vals,
-                   uint64_t cur_pc, uint64_t cur_offs, bool instrs_are_separate,
-                   DR_PARAM_OUT bool *reached_end_of_memrefs,
-                   DR_PARAM_OUT bool *interrupted);
+    preempted_by_kernel_event(raw2trace_thread_data_t *tdata, uint64_t cur_pc,
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.h:1495]: s/preempted/interrupted/?

File: clients/drcachesim/common/trace_entry.h [hunk 19]
```diff
@@ -692,6 +692,17 @@ typedef enum {
      */
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/common/trace_entry.h:692]: If we bump the trace version let's mention that here.  By doing that I don't think we need to bump the DR API version.

File: api/docs/release.dox [hunk 20]
```diff
@@ -128,6 +128,14 @@ The changes between version \DR_VERSION and 11.0.0 include the following compati
 changes:
  - Added X64 Linux support to dr_create_memory_dump(). This API has the same
```
Comment (derekbruening) [inline api/docs/release.dox:129]: This one is in the wrong list: please move it to the non-compatibility-affecting changes list below on line 133.

File: api/docs/release.dox [hunk 21]
```diff
@@ -128,6 +128,14 @@ The changes between version \DR_VERSION and 11.0.0 include the following compati
...
    restriction as dr_suspend_all_other_threads_ex().
+ - Removed uncompleted instructions and the corresponding memrefs from drmemtraces.
+   Instructions which are fetched but not completed due to asynchronous signal or
+   fault are removed. A new marker
+   #dynamorio::drmemtrace::TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION was added to indicate an
+   uncompleted instruction was removed. The value of the marker is the encoding of the
+   removed instruction up to a pointer's length. Added
+   OFFLINE_FILE_VERSION_RETIRED_INSTRUCTIONS_ONLY file type for drmemtraces with
```
Comment (derekbruening) [inline api/docs/release.dox:137]: This is not a "file type" as that sounds like the offline_file_type_t bitfield.  You increased the trace version to this new version to indicate traces where these are removed.

File: clients/drcachesim/common/trace_entry.h [hunk 22]
```diff
@@ -699,7 +699,11 @@ typedef enum {
      * are removed from the trace.
      *
      * The marker value is the raw encoding bytes of the instruction up to the
-     * length of a pointer.
+     * length of a pointer. The encoding will be incomplete for instructions
+     * with long encodings. It is best-effort to help understand the sequence of
+     * generated code where encodings are not available online. The PC of this
```
Comment (derekbruening) [inline clients/drcachesim/common/trace_entry.h:704]: s/online/offline/

File: clients/drcachesim/common/trace_entry.h [hunk 23]
```diff
@@ -699,7 +699,11 @@ typedef enum {
...
+     * instruction is available in a subsequent TRACE_MARKER_TYPE_KERNEL_EVENT
```
Comment (derekbruening) [inline clients/drcachesim/common/trace_entry.h:705]: Add `#` to make a link.

File: api/docs/release.dox (hunk 21 above)
Comment (derekbruening) [inline api/docs/release.dox:137]: Add # to make a link.

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 24]
```diff
@@ -193,6 +193,8 @@ Some of the more important markers are:
 
 - #dynamorio::drmemtrace::TRACE_MARKER_TYPE_CORE_WAIT - This is inserted by the trace scheduler (see \ref sec_drcachesim_sched) during replay of a previously recorded schedule when one core gets too far ahead of another according to the recorded timestamps.  This is an artificial wait to keep the replay on track, as opposed to the natural idle time of #dynamorio::drmemtrace::TRACE_MARKER_TYPE_CORE_IDLE.
 
+- #dynamorio::drmemtrace::TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION - This is inserted when an uncompleted instruction and its corresponding memory records are removed. The marker value contains the encoding of the removed instruction up to the length of a pointer. The encoding will be incomplete for instructions with long encodings. It is best-effort to help understand the sequence for generated code where encodings are not available offline. The PC of this instruction is available in a subsequent TRACE_MARKER_TYPE_KERNEL_EVENT marker.
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:196]: Use # to make a link for the KERNEL_EVENT ref.

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 25]
```diff
@@ -792,6 +794,43 @@ untaken conditional branch:
       801377      601843:     1159769 ifetch       4 byte(s) @ 0x00007fc2c3aa5c72 48 83 c4 48          add    $0x48, %rsp
 \endcode
 
+Here is an example of an asynchronous signal interrupting an instruction:
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:797]: Make it clear this is not what someone will see in a real trace!  (At least not a new-version trace.). Otherwise this may be very confusing.

File: clients/drcachesim/docs/drcachesim.dox.in [hunk 26]
```diff
@@ -792,6 +794,43 @@ untaken conditional branch:
...
+
+\code
+
+    46914793    33950158:     3767811 ifetch       5 byte(s) @ 0x000055f52911b75e c4 42 f8 f5 e8       bzhi   %r8, %rax, %r13
+    46914794    33950159:     3767811 ifetch       5 byte(s) @ 0x000055f52911b763 c4 62 f8 f5 c3       bzhi   %rbx, %rax, %r8
+    46914795    33950160:     3767811 ifetch       4 byte(s) @ 0x000055f52911b768 48 c1 eb 33          shr    $0x33, %rbx
+    46914796    33950161:     3767811 ifetch       3 byte(s) @ 0x000055f52911b76c 48 01 f3             add    %rsi, %rbx
+    46914797    33950162:     3767811 ifetch       5 byte(s) @ 0x000055f52911b76f b8 04 00 00 00       mov    $0x00000004, %eax
+    46914798    33950163:     3767811 ifetch       6 byte(s) @ 0x000055f52911b774 89 85 68 ff ff ff    mov    %eax, -0x98(%rbp)
```
Comment (derekbruening) [inline clients/drcachesim/docs/drcachesim.dox.in:806]: Also mark this line  somehow so it's clear this is here for demonstration purposes only and won't appear in a real trace?  Put it in parentheses??

File: clients/drcachesim/tools/view.cpp [hunk 27]
```diff
@@ -494,6 +494,8 @@ view_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
                       << " bytes>\n";
             break;
         case TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION:
+            // The value stores the encoding of the uncompleted instruction up
+            // to the length of a pointer. The encoding may not be complete.
```
Comment (derekbruening) [inline clients/drcachesim/tools/view.cpp:498]: "so we do not try to print it in the regular encoding style of the ISA".

File: clients/drcachesim/common/trace_entry.h [hunk 28]
```diff
@@ -962,7 +962,8 @@ typedef enum {
 #define OFFLINE_FILE_VERSION_HEADER_FIELDS_SWAP 5
 #define OFFLINE_FILE_VERSION_ENCODINGS 6
 #define OFFLINE_FILE_VERSION_XFER_ABS_PC 7
-#define OFFLINE_FILE_VERSION_RETIRED_INSTRUCTIONS_ONLY 8
+#define OFFLINE_FILE_VERSION_RETIRED_INSTRUCTIONS_ONLY \
+    8 /**< Trace version which has only retired instructions in drmemtraces.*/
 #define OFFLINE_FILE_VERSION OFFLINE_FILE_VERSION_RETIRED_INSTRUCTIONS_ONLY
```
Comment (derekbruening) [inline clients/drcachesim/common/trace_entry.h:967]: While at it, how about adding a doxygen comment for the current version too.

**Review by derekbruening — COMMENTED** [review]
Not seeing anything major, but could you resolve the 19 open comments?

**Comment by derekbruening** [comment]
> Not seeing anything major, but could you resolve the 19 open comments?

I didn't see replies: please reply when resolving so the replies show up in the email and it's clear what the resolution was (even if just "Done" that is helpful vs a silent resolution).

**Review by derekbruening — COMMENTED** [review]
Putting the top-level comment here to re-enable the re-request-review:

> Not seeing anything major, but could you resolve the 19 open comments?

I didn't see replies: please reply when resolving so the replies show up in the email and it's clear what the resolution was (even if just "Done" that is helpful vs a silent resolution).  Please reply to the 19 so it's clear what the current status is.

File: clients/drcachesim/common/trace_entry.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/common/trace_entry.h:692]: I've added a new trace version OFFLINE_FILE_VERSION_RETIRED_INSTRUCTIONS_ONLY.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 4 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.cpp:1873]: OK, they are reversed inside pre_read_buffer but then un-reversed here.

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp [hunk 29]
```diff
@@ -3156,6 +3156,248 @@ test_ifiltered(void *drcontext)
...
+            check_entry(entries, idx, TRACE_TYPE_FOOTER, -1);
+    }
+    {
+        std::cerr
+            << "\n===============\nTesting instr and memref removal by async signal\n";
+        instrlist_t *ilist = instrlist_create(drcontext);
+        // raw2trace doesn't like offsets of 0 so we shift with a nop.
+        instr_t *nop = XINST_CREATE_nop(drcontext);
+#ifdef AARCH64
+        // XXX i#5628: opnd_create_mem_instr is not supported yet on AArch64.
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/raw2trace_unit_tests.cpp:3221]: You're using a special function opnd_create_mem_instr() which gets the encoded address of an instruction into a load/store operand.  But you don't need that at all, right?  So no need for this #ifdef or anything: please just use a regular operand and remove the ifdef.

**Review by derekbruening — APPROVED** [review]
